---
title: "Transcriptomics and the analysis of RNA-Seq data"
author: "Linh-An Ho"
date: "2/20/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import today's data

The data for this hands-on session comes from a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

Glucocorticoids are used, for example, by people with asthma to reduce inflammation of the airways. The anti-inflammatory effects on airway smooth muscle (ASM) cells has been known for some time but the underlying molecular mechanisms are unclear.

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <- read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```

How many genes are we dealing with?
```{r}
nrow(counts)
```

We want to calculate the average counts per gene for control samples and for the drug treated samples. We will then compare these values to see if there is a difference overall between control and treated samples for a given gene.

This is our `metadata` metadata info that tells me where our control and treated samples are (i.e.which columns in `counts`)
```{r}
metadata
```

Now I want to access the `control` only rows in the metadata.
```{r}
#Shows the name of the control samples
control <- metadata[metadata$dex == "control",]
control
#metadata[metadata$dex == "control"]$id
```

Now use the $id of the control data to access the `control` columns in `counts` and calculate the average.
```{r}
control.mean <- rowSums( counts[, control$id])/ nrow(control)
names(control.mean) <- counts$ensgene
```


Now we find the average for the treated data samples
```{r}
#Shows the name of the control samples
treated <- metadata[metadata$dex == "treated",]
treated.mean <- rowSums( counts[, treated$id])/ nrow(treated)
names(treated.mean) <- counts$ensgene
```

We will combine our meancount data for bookkeeping purposes.
```{r}
meancounts <- data.frame(control.mean, treated.mean)
meancounts
```

Plot the data
```{r}
plot(meancounts)
```
Let's make this into a log-log plot
```{r}
plot.default(meancounts, log = "xy")
```

Here we calculate log2foldchange, add it to our meancounts data.frame and inspect the results either with the head() or the View() function for example.
```{r}
meancounts$log2fc <- log2(meancounts[, "treated.mean"]/meancounts[, "control.mean"])
head(meancounts)
```

Remove any gene that has a zero average counts.
```{r}
#`which` tells you which position rings true/false
#arr.ind tells you which row&col has a 0
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

How many genes do I have left?
```{r}
nrow(mycounts)
```


A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Letâ€™s filter the dataset both ways to see how many genes are up or down-regulated.
```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

How many genes are up or down regulated?
```{r}
sum(up.ind)
sum(down.ind)
```

## Do this properly with DESeq2
```{r}
library(DESeq2)
```

Set up input for DESeq by specifying countdata, metadata, and the experiment design (i.e. which comparisons we are interested in)
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

```{r}
dds <- DESeq(dds)
```
```{r}
res <- results(dds)
res
```

```{r}
# plotting values of fold change with the stats
plot(res$log2FoldChange, res$padj)
```

```{r}
plot(res$log2FoldChange, log(res$padj))
```

Let's make a plot the way the rest of the world does
```{r}
plot( res$log2FoldChange,  -log(res$padj))
```


Here we first make a volcano plot with base graphics and color by our res$sig+1 (we add 1 so we dont have 0 as a color and end up with white points):
```{r}
res$sig <- res$padj<0.05 & abs(res$log2FoldChange)>2

# How many of each?
table(res$sig)

# Set the color palette for our plot
palette( c("gray","blue") )

plot( res$log2FoldChange,  -log(res$padj), 
 col=res$sig+1, ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.1), col="darkgray", lty=2)
```

We could also setup a custom color vector indicating transcripts with large fold change and significant differences between conditions:

```{r}
# Setup our custom point color vector (to distinguish p-value)
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

#Volcano plot with custom colors (0.01 and 2 are the bounds for statistical significance)
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"
plot(res$log2FoldChange, -log(res$padj),
      col=mycols, 
     ylab="-Log(P-value)", 
     xlab="Log2(FoldChange)" )

abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```


Make a ggplot
```{r}
library(ggplot2)
ggplot(as.data.frame(res), aes(log2FoldChange, -log10(pvalue), col=sig)) + 
    geom_point() + 
    ggtitle("Volcano plot")
```

