---
title: "Pathway Analysis from RNA-Seq"
author: "Linh-An Ho"
date: "2/27/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import RNA-Seq counts and metadata

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703
The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.

Here we read the two input files
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"
```

```{r}
# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

There are `r nrow(countData)` genes in this dataset and `r ncol(countData)` experiments (samples)...

Looks like ther is an extra 'length' column in our countData that we need to remove
```{r}

# Note we need to remove the odd first $length col (use -1)
countData <- as.matrix(countData[,-1])
head(countData)
```


A little bit more tidying work required. Let's remove the genes where we have no data (i.e. zero counts in all experiments)

Let's take the rowSums() and exclude the rows with 0 values
```{r}
count.data <- countData[rowSums(countData) != 0,]
```

Check how many genes we have left
```{r}
nrow(count.data)
```

## Running DESeq2

Now we get to run DESeq on our data.
```{r}
library(DESeq2)
dds <-  DESeqDataSetFromMatrix(countData=count.data,
                             colData=colData,
                             design=~condition)
dds <-  DESeq(dds)
```

Let's see our results
```{r}
res <-  results(dds)
res
```

## Volcano Plot

Summary plot of our results to date
```{r}
plot( res$log2FoldChange, -log(res$padj) )
abline(h= -log(0.05), col = "red")
abline(v=c(-2,2), col = "red")
```

Since we mapped and counted against the Ensembl annotation, our results only have information about Ensembl gene IDs. However, our pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. So lets add them as we did the last day.

I would install these with `BiocManager::install("AnnotationDbi")` and `BiocManager::install("org.Hs.eg.db")`
```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

# What is available in the org.Hs.eg.db dataset?
columns(org.Hs.eg.db)
```

Let's add extra annotation to our results object.
```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

head(res)
```

Now that we have our geneIDs in the correct format, we can do pathway analysis

## Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.


First I need to install the packages I need 
```{r}
#BiocManager::install( c("pathview", "gage", "gageData") )
library(pathview)
library(gage)
library(gageData)
```


```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Make an input vector of our number quantity of interest (i.e. fold change values and make sure it has ENTREZ gene names associated with it)
```{r}
# The entrez IDs are needed to be read for KEGG
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now run **gage** with this input vector nad our KEGG genesets
```{r}
# Get the results
keggres <-  gage(foldchanges, gsets=kegg.sets.hs)
```

What is in our results? We can use the `attributes()` function to see what is returned.
```{r}
attributes(keggres)
```

We use the KEGG pathway ID from our top hit along with our `foldrange` vector to view our results as a more useful figure.
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```


```{r}
knitr::include_graphics(hsa04110.pathview.png)
```

